[Backend] recommendation

furkannane opened the pull request:
## Type of change

- [X] New feature (non-breaking change which adds functionality)


# Description

I implemented the GET recommendation endpoint which returns some products to recommend to the customers according to their purchases. 


# Test results

1 test is added for the recommended products and it is passed.

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have checked my code and corrected any misspellings
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

furkannane committed:
Recommendation Function is created

Recommendation function is created

furkannane committed:
Url is added and init.py import added

Url is added and init.py import added

furkannane committed:
tests are added and while adding tests, some corrections are made as well.

tests are added and while adding tests, some corrections are made as well

omerfdeniz commented on ecommerce/backend/API/views/recommendation.py:
```diff
@@ -0,0 +1,81 @@
+
+            recommended_products.append(serialized_product)
+
+    return Response({'products':recommended_products})
```
add status and message to the response please:

```suggestion
    return Response({'status': {'successful': True, 'message': "Successfully retrieved the recommendations."}, 'products':recommended_products})
```

omerfdeniz commented on ecommerce/backend/API/tests/test_recommendation.py:
```diff
@@ -0,0 +1,76 @@
+    def test_1_recommendation(self):
+        response = self.client.get(reverse(recommend_products))
+
+        self.assertEqual(len(response.data["products"]), 2)
```
please also chech for 200 and status in the response:
```suggestion
        self.assertEqual(len(response.data["products"]), 2)
        self.assertEqual(response.status_code, 200)
        self.assertEqual(response.data["status"]["successful"], True)
```

furkannane replied:
Done

omerfdeniz commented on ecommerce/backend/API/views/recommendation.py:
```diff
@@ -0,0 +1,81 @@
+        # get 5 products that are non-deleted 
+        products = Product.objects.filter(is_deleted=False)[:5]
+
+        for product in products:
+            serializer = ProductResponseSerializer(product)
+            serialized_product = serializer.data
+
+            recommended_products.append(serialized_product)
+
+    return Response({'products':recommended_products})
```
It is better practice to use serializer with many field for a list serialization. Also please add the status and message to the response
```suggestion
        product_serializer = ProductResponseSerializer(products, many=True)
        return Response({'status': {'successful': True, 'message': "Successfully retrieved the recommendations."}, 'products': product_serializer.data})
        
```

furkannane replied:
Added status and message.

omerfdeniz reviewed (changes requested):
Address the response type and serialization please

bekir96 commented on ecommerce/backend/getflix/urls.py:
```diff
@@ -49,5 +49,6 @@
     path('customer/orders', views.order.customer_order),
     path('messages', views.message.manage_messages),
     path('notifications', views.notification.manage_notifications, name="notifications"),
-    path('lists', views.product_list.product_list_create)
+    path('lists', views.product_list.product_list_create),
+    path('recommendation', views.recommendation.recommend_products)
 ]
```
can you resolve conflict in urls.py

bekir96 reviewed (changes requested):
The master may have changed, you will have to look at conflicts.

umutoksuz commented on ecommerce/backend/API/views/recommendation.py:
```diff
@@ -0,0 +1,81 @@
+    # get user
+    user = request.user
+    user_id = request.user.id
+
```
Here if no token is provided in request header, and user will be none. You should check for that and return an error in case

umutoksuz reviewed (changes requested):
Some changes requrested

furkannane committed:
Test is updated according to feedback

Test is updated according to feedback: now also checking for status code and is status successful.

furkannane committed:
Added status field to the response according to the feedback

Added status field to the response according to the feedback

furkannane committed:
Added check for valid user according to the feedback

Added check for valid user according to the feedback

furkannane committed:
Merge branch 'master' into B-Recommendation
